modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -510,8 +510,8 @@ def map_func(left, right=right, kwargs=kwargs):
 
     # Reindex/reset_index (may shuffle data)
     def reindex(self, axis, labels, **kwargs):
-        new_index = self.index if axis else labels
-        new_columns = labels if axis else self.columns
+        new_index, _ = (self.index, None) if axis else self.index.reindex(labels)
+        new_columns, _ = self.columns.reindex(labels) if axis else (self.columns, None)
```

Comment:
The main change is that if you perform reindexing on `self.index`, then the metadata does not disappear, as with a simple new index assignment.

Reply:
This reindexing would block the flow for a couple of seconds in case of a long axis (10 mils is enough to take it up to 1 second)
<details><summary>repro</summary>

```python
import pandas
import numpy as np
from timeit import default_timer as timer

index = pandas.Index(np.arange(10_000_000))
reindexer = np.arange(10_000_000)
np.random.shuffle(reindexer)

t1 = timer()
index = index.reindex(reindexer)
print(f"reindexing: {timer() - t1}")
# Output:
# reindexing: 1.0896086000000003
```

</details>

I believe there's a know fixed set of metadata attributes that can be explicitly copied from `self.index` to `labels`
```suggestion
        new_index = self.index if axis else copy_index_metadata(from=self.index, to=labels)
        new_columns = copy_index_metadata(from=self.columns, to=labels) if axis else self.columns
```